### PR TITLE
[SPARK-18420][BUILD] Fix the errors caused by lint check in Java

### PR DESCRIPTION
**Pull request by ConeyLiu**
## What changes were proposed in this pull request?

Small fix, fix the errors caused by lint check in Java

- Clear unused objects and `UnusedImports`.
- Add comments around the method `finalize` of `NioBufferedFileInputStream`to turn off checkstyle.
- Cut the line which is longer than 100 characters into two lines.

## How was this patch tested?
Travis CI.
```
$ build/mvn -T 4 -q -DskipTests -Pyarn -Phadoop-2.3 -Pkinesis-asl -Phive -Phive-thriftserver install
$ dev/lint-java
```
Before:
```
Checkstyle checks failed at following occurrences:
[ERROR] src/main/java/org/apache/spark/network/util/TransportConf.java:[21,8] (imports) UnusedImports: Unused import - org.apache.commons.crypto.cipher.CryptoCipherFactory.
[ERROR] src/test/java/org/apache/spark/network/sasl/SparkSaslSuite.java:[516,5] (modifier) RedundantModifier: Redundant 'public' modifier.
[ERROR] src/main/java/org/apache/spark/io/NioBufferedFileInputStream.java:[133] (coding) NoFinalizer: Avoid using finalizer method.
[ERROR] src/main/java/org/apache/spark/sql/catalyst/expressions/UnsafeMapData.java:[71] (sizes) LineLength: Line is longer than 100 characters (found 113).
[ERROR] src/main/java/org/apache/spark/sql/catalyst/expressions/UnsafeArrayData.java:[112] (sizes) LineLength: Line is longer than 100 characters (found 110).
[ERROR] src/test/java/org/apache/spark/sql/catalyst/expressions/HiveHasherSuite.java:[31,17] (modifier) ModifierOrder: 'static' modifier out of order with the JLS suggestions.
[ERROR]src/main/java/org/apache/spark/examples/ml/JavaLogisticRegressionWithElasticNetExample.java:[64] (sizes) LineLength: Line is longer than 100 characters (found 103).
[ERROR] src/main/java/org/apache/spark/examples/ml/JavaInteractionExample.java:[22,8] (imports) UnusedImports: Unused import - org.apache.spark.ml.linalg.Vectors.
[ERROR] src/main/java/org/apache/spark/examples/ml/JavaInteractionExample.java:[51] (regexp) RegexpSingleline: No trailing whitespace allowed.
```

After:
```
$ build/mvn -T 4 -q -DskipTests -Pyarn -Phadoop-2.3 -Pkinesis-asl -Phive -Phive-thriftserver install
$ dev/lint-java
Using `mvn` from path: /home/travis/build/ConeyLiu/spark/build/apache-maven-3.3.9/bin/mvn
Checkstyle checks passed.
```

### CODE REVIEW

**Comment by HyukjinKwon** [comment]
Hi @ConeyLiu, It seems 

```
[ERROR] src/main/java/org/apache/spark/io/NioBufferedFileInputStream.java:[133] (coding) NoFinalizer: Avoid using finalizer method.
```

is missed. If it is a required change, I guess we should disable the lint check for the lines.

**Comment by ConeyLiu** [comment]
@HyukjinKwon I'm sorry for that, I've fixed it, please take a look. Thanks a lot.

File: core/src/main/java/org/apache/spark/io/NioBufferedFileInputStream.java [hunk 1]
```diff
@@ -129,9 +129,4 @@ public synchronized void close() throws IOException {
     fileChannel.close();
     StorageUtils.dispose(byteBuffer);
   }
-
-  @Override
-  protected void finalize() throws IOException {
```
Comment (srowen) [inline core/src/main/java/org/apache/spark/io/NioBufferedFileInputStream.java]: Why was this removed? it needs to be here for sure

File: sql/catalyst/src/test/java/org/apache/spark/sql/catalyst/expressions/HiveHasherSuite.java [hunk 2]
```diff
@@ -28,7 +28,7 @@
 import java.util.Set;
 
 public class HiveHasherSuite {
-  private final static HiveHasher hasher = new HiveHasher();
+  private final HiveHasher hasher = new HiveHasher();
```
Comment (srowen) [inline sql/catalyst/src/test/java/org/apache/spark/sql/catalyst/expressions/HiveHasherSuite.java]: You made this non-static; that's not a style issue right? it may be OK but some of these changes don't seem to match the description

**Comment by srowen** [comment]
They're caused by lint right? not checkstyle? and these are only Java, because that's all that's not automatically tested. Please update the title/description with more meaningful details, including the JIRA.

**Review by srowen — CHANGES_REQUESTED** [review]

**Comment by ConeyLiu** [comment]
@srowen  Thanks for the review, I've updated it.

**Comment by HyukjinKwon** [comment]
@ConeyLiu Let us please make the title `[SPARK-18420][BUILD] Fix the errors caused by lint check in Java` for this PR and Jira you opened.

I think for the one below:

```
[ERROR] src/main/java/org/apache/spark/io/NioBufferedFileInputStream.java:[133] (coding) NoFinalizer: Avoid using finalizer method.
```

we should disable the lint for the lines because it seems a required one. I guess you should disable the lines or work around this.

Could we then remove `hasher` in `HiveHasherSuite.java` if it is not used? FWIW, I just checked that this private instance is not referenced in `HiveHasherSuite.java` for sure.

**Comment by ConeyLiu** [comment]
@HyukjinKwon Thanks for the review and suggestion, I've updated it. Clear the unused object `hasher` and add suppression rules for the method `finalize` of `NioBufferedFileInputStream`. Please take a look, thanks a lot.

File: dev/checkstyle-suppressions.xml [hunk 3]
```diff
@@ -30,6 +30,8 @@
 <suppressions>
     <suppress checks=".*"
               files="core/src/main/java/org/apache/spark/util/collection/TimSort.java"/>
+    <suppress checks="NoFinalizer"
+              files="core/src/main/java/org/apache/spark/io/NioBufferedFileInputStream.java"/>
```
Comment (HyukjinKwon) [inline dev/checkstyle-suppressions.xml]: Ah, I thought we could disable it only for some lines. I am fine with this if it is only the way.

Comment (ConeyLiu) [inline dev/checkstyle-suppressions.xml]: @HyukjinKwon  Also we could try `// scalastyle:off ...  // scalastyle:on`. I'll test this way.

**Review by srowen — APPROVED** [review]

File: sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/UnsafeArrayData.java [hunk 4]
```diff
@@ -109,7 +109,8 @@ public void pointTo(Object baseObject, long baseOffset, int sizeInBytes) {
     // Read the number of elements from the first 8 bytes.
     final long numElements = Platform.getLong(baseObject, baseOffset);
     assert numElements >= 0 : "numElements (" + numElements + ") should >= 0";
-    assert numElements <= Integer.MAX_VALUE : "numElements (" + numElements + ") should <= Integer.MAX_VALUE";
+    assert numElements <= Integer.MAX_VALUE :
+            "numElements (" + numElements + ") should <= Integer.MAX_VALUE";
```
Comment (HyukjinKwon) [inline sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/UnsafeArrayData.java]: Hi @ConeyLiu, it is a super minor but it seems we should make the indentation for this (and the same instances here) double-spaced. I at least could find some examples for this such as in `UnsafeMemoryAllocator`

```java
    assert (memory.obj == null) :
      "baseObject not null; are you trying to use the off-heap allocator to free on-heap memory?";
```

or in `ShuffleSortDataFormat`

```java
    assert (length <= buffer.size()) :
      "the buffer is smaller than required: " + buffer.size() + " < " + length;
```

File: examples/src/main/java/org/apache/spark/examples/ml/JavaInteractionExample.java [hunk 5]
```diff
@@ -48,8 +47,7 @@ public static void main(String[] args) {
       RowFactory.create(5, 9, 2, 7, 10, 7, 3),
       RowFactory.create(6, 1, 1, 4, 2, 8, 4)
     );
-
```
Comment (HyukjinKwon) [inline examples/src/main/java/org/apache/spark/examples/ml/JavaInteractionExample.java]: If you are supposed to push more commits, let us revert this extra newline removal. I guess some guys don't like such changes. At least I was told twice before.

**Comment by HyukjinKwon** [comment]
For sure, I ran 

```bash
$ ./dev/lint-java
Using `mvn` from path: .../mvn
Checkstyle checks passed.
```

It seems fine. It looks good to me for few minor comments above.

**Comment by ConeyLiu** [comment]
@HyukjinKwon  I'm sorry for so late reply. I update the code and style followed by your suggestion, that's so useful. And also I change the method of disabling the check around `finalize`.  Please take a loot or give some suggestion.  Thanks a lot.

File: dev/checkstyle.xml [hunk 6]
```diff
@@ -52,6 +52,20 @@
...
+      // checkstyle:off no.XXX (such as checkstyle.off: NoFinalizer)
+      ...  // stuff that breaks the styles
+      // checkstyle:on
+    -->
+    <module name="SuppressionCommentFilter">
+        <property name="offCommentFormat" value="checkstyle.off\: ([\w\|]+)"/>
+        <property name="onCommentFormat" value="checkstyle.on\: ([\w\|]+)"/>
+        <property name="checkFormat" value="$1"/>
+    </module>
+
```
Comment (HyukjinKwon) [inline dev/checkstyle.xml]: Ah, I am sorry @ConeyLiu. I should have taken a look ahead first. I actually rather meant agreeing the previous change if we should introduce such new property... Although I (as just one of contributors) think this might be worth, maybe this is orthogonal with what the PR tries to fix.

Maybe, I think we could keep this change and then propose a new separate PR later if other people also agree with this.

Comment (ConeyLiu) [inline dev/checkstyle.xml]: Ok, maybe we need more advices.

**Comment by ConeyLiu** [comment]
@srowen hi owen, Can you give me some advice, and if so, what a great thing.

File: dev/checkstyle.xml [hunk 7]
```diff
@@ -168,5 +182,6 @@
         <module name="UnusedImports"/>
         <module name="RedundantImport"/>
         <module name="RedundantModifier"/>
+        <module name="FileContentsHolder"/>
```
Comment (srowen) [inline dev/checkstyle.xml]: What is this needed for, the `checkstyle:off` mechanism?

Comment (ConeyLiu) [inline dev/checkstyle.xml]: @srowen `SuppressionCommentFilter`and `FileContentsHolder` must work together,http://stackoverflow.com/questions/5761188/checkstyle-suppressioncommentfilter-not-ignoring-specified-rule

**Review by srowen — APPROVED** [review]

**Comment by HyukjinKwon** [comment]
I just checked for sure.

```
$ ./dev/lint-java
Using `mvn` from path: .../mvn
Checkstyle checks passed.
```

**Comment by ConeyLiu** [comment]
@HyukjinKwon @srowen Thank you very much.

**Comment by srowen** [comment]
Merged to master/2.1
